### PR TITLE
Pinmux and top generation

### PR DESCRIPTION
**Pull request by marnovandermaas**
This implements the pinmux. It generates the number of GPIO, UART, I2C and SPI blocks specified by the `data/blocks.yml`. There is a `util/top_gen.py` script that uses templates to generate the appropriate changes to the crossbar and the package that specifies the number of each block.

This PR also does the following useful things:
- Adds output enables to the GPIO module.
- Make it so that multiple GPIO registers can live in the same 4 KiB memory.
- Reduced number of SPI blocks from 7 to 5.

There are a number of possible refinements to this design which can be addressed as follow-up PRs and I have captured them here: https://github.com/lowRISC/sonata-system/issues/169

Closes: #148

### CODE REVIEW

File: sw/legacy/common/uart.c
Comment (GregAC) [inline sw/legacy/common/uart.c:13]: Code cleanup here isn't relevant to the rest of the commit

File: sw/legacy/demo/ethernet/ksz8851.c
Comment (GregAC) [inline sw/legacy/demo/ethernet/ksz8851.c:225]: Is this a tab to space change? (Or vice versa hard to tell in the github UI...)

File: data/blocks.yml
Comment (GregAC) [inline data/blocks.yml:19]: With the pinmux we should seek to reduce the number of SPI and UART instances, we've got so many as we've needed one per physical interface.

Need to think carefully about i2c as well, it's a big block so good resource savings by reducing the number. Though still if we're not using the resources for anything else may as well have more i2c, 2 feels like a reasonable number though we could likely manage with 1.

File: data/pins.yml
Comment (GregAC) [inline data/pins.yml:103]: This setup is fine for getting things going but for the v1.0 release I do want to enable more flexibility, e.g. so a given set of rpi HAT pins could be GPIO, could be I2C, could be UART or could be SPI. If you restrict yourself to what the rpi can actually do (uncertain how flexible it is in connecting its own SPI etc blocks to the GPIO pins) you may have a more restricted set but it is also a general purpose header people can connect non RPI HAT stuff to so flexibility is useful.

File: util/top_gen.py
Comment (GregAC) [inline util/top_gen.py:9]: We should decouple the pinmux generation from the top gen and make sure validation/parsing functionality is nicely separated.

As we've deprioritised top gen I'd focus entirely on a standalone pinmux tool that's written with reasonable modularity so at a later stage it could be used within topgen rather than run as a separate invocation.

File: util/top_gen.py
Comment (GregAC) [inline util/top_gen.py:16]: We should use an existing YAML schema validation library, e.g. https://pypi.org/project/schema/ (I'm sure there are more there's another one I used in the past but cannot remember what it's called but schema looks pretty decent from a quick glance).

That said for now we should minimise validation work, just putting some asserts in places where you're using the data structure pulled from the YAML to check things are as you expect should do for now. We need to prioritise getting the pinmux generated and usable over good error checking etc.

**Review by GregAC — COMMENTED** [review]
Was there a reason to do the auto-generated top? I guess it's only a modest amount of work in this PR  but we had de-prioritised the topgen/system builder work for the v1.0 release.

File: sw/legacy/demo/ethernet/ksz8851.c
Comment (marnovandermaas) [reply sw/legacy/demo/ethernet/ksz8851.c:225]: This removes trailing whitespace.

File: util/top_gen.py
Comment (marnovandermaas) [reply util/top_gen.py]: Using Pydantic instead. Thanks to @HU90m for the work.

File: sw/legacy/common/uart.c
Comment (marnovandermaas) [reply sw/legacy/common/uart.c:13]: Moved this to a cleanup commit.

File: sw/legacy/demo/ethernet/ksz8851.c
Comment (marnovandermaas) [reply sw/legacy/demo/ethernet/ksz8851.c:225]: Moved this to a cleanup commit with message saying what's done.

File: data/pins.yml
Comment (marnovandermaas) [reply data/pins.yml]: I've captured this here: https://github.com/lowRISC/sonata-system/issues/169

File: util/top_gen.py
Comment (marnovandermaas) [reply util/top_gen.py]: I've captured this here: https://github.com/lowRISC/sonata-system/issues/169

File: data/blocks.yml
Comment (marnovandermaas) [reply data/blocks.yml]: I've captured this here: https://github.com/lowRISC/sonata-system/issues/169

File: rtl/system/pinmux.sv
Comment (GregAC) [inline rtl/system/pinmux.sv:124]: I feel like it'd be better to use reggen to generate all of this. It is a pain but it keeps things consistent with the other IPs. We've already got an issue in Sonata that we've got some IPs using the legacy bus from Ibex demo system each with their own custom register logic, this adds another kind of custom register logic. Long term it's better if everything is using the same methods consistently (i.e. reggen for everything)l

So pinmux generator would also produce a hjson that can then be fed through reggen. This probably means we need to vendor reggen into this repository and patch it so it doesn't generate the integrity related stuff. As the SPI block uses reggen and currently the reggen output has been manually altered to remove the integrity stuff it is something we want anyway. Especially as we should also move the legacy IPs on to reggen to get consistency.

**Review by GregAC — COMMENTED** [review]
I appreciate this need to be merged to avoid having to keep rebasing it and avoiding conflicts and not all comments from my review need to be addressed but I do view the use of Zs within pinmux and the combination of pinmux/pad functionality as critical things that must be addressed before we can merge.

I would like to get the reggen stuff sorted before v1.0 as well as the updated GPIO module (where we can have a single GPIO block with multiple sets of registers to control all GPIO pins that can live on a single port on the crossbar).

File: rtl/system/sonata_system.sv
Comment (marnovandermaas) [reply rtl/system/sonata_system.sv]: Ok, I have now removed this HalfWordWidth in the new changes.

File: data/xbar_main.hjson
Comment (marnovandermaas) [reply data/xbar_main.hjson]: I've now removed these extra GPIO modules and implemented a gpio_array module that instantiates multiple GPIO register groups in the same address space.

File: data/top_config.toml
Comment (marnovandermaas) [reply data/top_config.toml]: Ok, it was part of the GPIO block but I've now changed the syntax to make it more clear.

File: rtl/system/sonata_system.sv
Comment (marnovandermaas) [reply rtl/system/sonata_system.sv:41]: This is now done with the new IP block.

File: rtl/fpga/top_sonata.sv
Comment (marnovandermaas) [reply rtl/fpga/top_sonata.sv]: Yes, I've removed this now based on the new GPIO unit with enables.

File: rtl/templates/pinmux.sv.tpl
Comment (marnovandermaas) [reply rtl/templates/pinmux.sv.tpl]: Agreed, this also allows the pinmux to be built using Verilator which is nice.

File: rtl/templates/pinmux.sv.tpl
Comment (marnovandermaas) [reply rtl/templates/pinmux.sv.tpl:20]: I agree I've essentially merged the padring and the pinmux. I'm not sure if splitting it up will make a cleaner design, but I'm not against it. Would it be ok to merge the PR as is and then potentially split the padring and the pinmux in a future PR?

File: rtl/system/pinmux.sv
Comment (marnovandermaas) [reply rtl/system/pinmux.sv:126]: I'm not against having a way to generate registers, although Hugo mentioned an alternative SystemRDL as an alternative. Either way I think this is something we can resolve in a follow-up PR.

Comment (marnovandermaas) [reply rtl/system/pinmux.sv:126]: It's also mentioned as a task in the pinmux refinement issue: https://github.com/lowRISC/sonata-system/issues/169

File: rtl/templates/pinmux.sv.tpl
Comment (marnovandermaas) [reply rtl/templates/pinmux.sv.tpl:20]: It's also mentioned in the pinmux refinement issue: https://github.com/lowRISC/sonata-system/issues/169

File: rtl/system/pinmux.sv
Comment (GregAC) [reply rtl/system/pinmux.sv:126]: Ultimately I think we should have consistency. Right now the OpenTitan reggen is probably the quickest way to consistency. I don't want to introduce another register interface generator so if we're going to use SystemRDL we'd need to switch over wholesale which isn't something we can support right now.

File: sw/cheri/checks/cheri_sanity.cc
Comment (HU90m) [inline sw/cheri/checks/cheri_sanity.cc:24]: Remember to open an issue for this

File: rtl/system/pinmux.sv
Comment (GregAC) [reply rtl/system/pinmux.sv:7950]: Oh and this makes more sense as a bit-wise AND I think

File: rtl/templates/pinmux.sv.tpl
Comment (GregAC) [reply rtl/templates/pinmux.sv.tpl:20]: It's about separation of concerns. The functionality of a pinmux is distinct from a pad instance/pad ring so it's better to de-couple them. It's particularly critical in this instance for two reasons:

1. Use of `inout` and `Z`. As I said in my earlier comment these need to be carefully handled in RTL. The pinmux is a large auto-generated module so there's lot to look at and think about so don't want to add ensuring correct usage of tristate as part of this. By keeping the pinmux strictly digital logic and isolating the tristate stuff into a per pad module there's a far smaller bit of RTL to review for correct tristate usage and lower maintenance burden going forward (we only have to think about tristate when we touch the pad instance). It would have reduced the effort required on my part to review this for instance.

2. Ease of switching pad implementation. Different targets (different FPGAs, ASIC etc) may want different pad instances. The generic `assign pad = pad_oe ? pad_o : 'z`  does do the job for FPGA most of the time but there may be reasons you want a technology specific pad instance and it would be essential were this ever to be used against an ASIC target.

Happy for this to get merged without this getting changed but I don't want it dropped as a 'nice-to-have' task that doesn't get done.

**Review by GregAC — APPROVED** [review]
Thanks for your work on this @marnovandermaas I think we're good to go once the issue with the I2C signal combine has been addressed.

File: doc/ip/pinmux.md
Comment (HU90m) [inline doc/ip/pinmux.md:9]: Once this is in and settled, we should document the default configuration. Mention of other top configurations should probably be left out of the reference manual and be somewhere else, somewhat hidden from normal users.

Comment (marnovandermaas) [reply doc/ip/pinmux.md]: The table for this is automatically generated now and it mentions the default selected value is `'b10`

File: rtl/system/pinmux.sv
Comment (marnovandermaas) [reply rtl/system/pinmux.sv]: Good spot! I've added the selectors in here now and changed to bitwise operators.

File: sw/cheri/checks/cheri_sanity.cc
Comment (marnovandermaas) [reply sw/cheri/checks/cheri_sanity.cc:24]: Here you go:
https://github.com/lowRISC/sonata-system/issues/206

File: rtl/system/pinmux.sv
Comment (marnovandermaas) [reply rtl/system/pinmux.sv:14]: Captured in: https://github.com/lowRISC/sonata-system/issues/169

**Pull request merged**
